src/main/xerus/monstercat/api/Player.kt

```diff
@@ -136,12 +143,15 @@ object Player : FadingHBox(true, targetHeight = 25) {
 			add(closeButton)
 		}
 	}
-	
+
 	private fun setVolume() {
 		player?.volume = volumeSlider.value.square
 	}
-	
+
 	private var player: MediaPlayer? = null
+	// Listeners are notified when the MediaPlayer is swapped.
+	var playerListeners = mutableSetOf<() -> Unit>()
```

Comment:
Sets sind unglaublich ineffizient in Java, sie verwenden darunter eine Hashmap.
Meiner Meinung nach sollte das hier so aussehen: 
```kotlin
var playerListeners = ArrayList<(MediaPlayer?) -> Unit>()
```
wobei null gegeben wird wenn der aktuelle MediaPlayer disposed wurde und ansonsten der neue MediaPlayer

Reply:
Sets sind zwar nicht so speichereffizient wie ein Array da drunter, haben dafür aber fast nur O(1) Operationen... daher tendiere ich normalerweise dazu, Sets zu nutzen, wenn ich keine List brauche.

In diesem Fall nutze ich die Listeners ja eigentlich nur zum einfügen und iterieren, d.h. eine List sollte genauso gut sein.

In Kotlin würde ich zudem eher

`mutableListOf<(MediaPlayer?) -> Unit>()`

verwenden, damit klar ist, dass die Liste mutiert wird (damit auch nicht thread-safe ist, was wiederum ein Problem darstellen könnte).

Reply:
1. Eine ArrayList ist immer mutable. Bei `mutableListOf` hast du keine Garantie woran du bist, aber wahrscheinlich gibt es auch einfach eine ArrayList zurück.
2. Das sollte natürlich eine `val` sein, also final.

Reply:
Richtig, ist hauptsächlich der Semantik halber. Persönlich halte ich nicht so viel von Mikrooptimierungen an Stellen, wo sie höchstwahrscheinlich keinen Unterschied machen. Da die Kotlin Standardbibliothek bereits diese praktischen Initializer-Funktionen bietet, schlage ich vor sie einfach zu nutzen. :)

Reply:
`val` ist keine Mikrooptimierung, sondern eine klare allseitige Style-Empfehlung um den Code weniger fehleranfällig zu machen. Deswegen ist es in Kotlin auch viel einfacher eine finale variable zu deklarieren (einfach l statt r), im Gegensatz zu Java, wo man dafür ein extra keyword braucht.

Bezüglich der Initializer-Funktion verstehe ich nicht so ganz, was daran so besonders ist? Ich habe schon immer meine listen direkt als ArrayList deklariert und sehe da jetzt keinen praktischen Unterschied. Es ist ja nichtmal kürzer ^^